### PR TITLE
Fix construtor for spec const matrix and vector

### PR DESCRIPTION
**Pull request by Qining**
Fix issue: #237

1. The code generated for matrix constructor should 1) build column
vectors first, 2) build matrix with the vectors.

2. When there is only one scalar type constituent in vector's
constructor, we should populate the constituent to fill all the slots in
the vector. As for matrix, the single constituent should be populated to
the diagonal positions (top-left to bottom-right diagonal).

remove createSpvConstantFromConstSubTree()

### CODE REVIEW

File: Test/spv.specConstantComposite.vert [hunk 1]
```diff
@@ -42,6 +42,11 @@ const vec4 spec_vec4_all_spec =
     vec4(spec_float, spec_float, spec_float, spec_float);
 const vec4 spec_vec4_partial_spec =
     vec4(spec_float, spec_float, 300.14, 300.14);
+const vec4 spec_vec4_from_one_scalar = vec4(spec_float);
+
+// Matrix constructor
+const mat2x3 spec_mat2x3 = mat2x3(spec_float, spec_float, spec_float, 1.1, 2.2, 3.3);
+const mat2x3 spec_mat2x3_from_one_scalar = mat2x3(spec_float);
```
Comment (Qining) [inline Test/spv.specConstantComposite.vert]: There are some changes in the test base results that are not caused by the added spec constants here. Such changes are only mismatches of instruction orders due to the remove of createSpvConstantFromConstSubTree().

In original createSpvConstantFromConstSubTree(), we call convertGlslangToSpvType() even though we don't use them immediately (actually the `typeId` is only used for handling aggregate nodes). But such a call will generate type declaration instruction, so removing the redundant call to convertGlslangToSpvType() will change the instruction order.

**Comment by Qining** [comment]
@johnkslang @da3k PTAL

**Comment by johnkslang** [comment]
Just to get better oriented first... we previously decided that the unary and binary nodes covered all the cases in KHR_vulkan_glsl, so I wonder if the fix here is to not be setting the spec-constness on these nodes in the front end?

(I'll have more time to look deeper, but want to make sure the basics are in place in the meantime.)

**Comment by Qining** [comment]
The problem is that the original createSpvConstantFromConstSubTree() handles aggregate nodes without using the `traverse(this)` (different with binary and unary nodes). And bug is caused by missing functionalities in there.

This CL actually removed the createSpvConstantFromConstSubTree() and unified the handlers of all kinds of nodes with traverse(this) then getAccessChainLoad().

So, in this CL, we use the same visit*() functions to handle both run-time instruction generation and spec constant op instruction generation.

**Comment by Qining** [comment]
Any suggestions for this CL?

**Comment by johnkslang** [comment]
I asked the wrong way.  I think your code changes are just fine for AST -> SPIR-V, and I'm just asking about which composites GLSL -> AST should be allowed to be created as spec-const nodes.

File: SPIRV/SpvBuilder.cpp [hunk 2]
```diff
@@ -1714,6 +1732,20 @@ Id Builder::createCompositeConstruct(Id typeId, std::vector<Id>& constituents)
 {
     assert(isAggregateType(typeId) || (getNumTypeConstituents(typeId) > 1 && getNumTypeConstituents(typeId) == (int)constituents.size()));
 
+    if (generatingOpCodeForSpecConst) {
+        // Sometime, even in spec-constant-op mode, the constant composite to be
+        // constructed may not be a specialization constant.
+        // e.g.:
+        //  const mat2 m2 = mat2(a_spec_const, a_front_end_const, another_front_end_const, third_front_end_const);
```
Comment (johnkslang) [inline SPIRV/SpvBuilder.cpp]: The AST nodes should be correct about what to create as a spec-const.  Is that not the case?

Comment (Qining) [inline SPIRV/SpvBuilder.cpp]: Yes, exactly!

**Comment by Qining** [comment]
I didn't see anything about "spec constant matrix" in KHR_glsl_vulkan. Actually I also feel it is suspicious, as at line 226, we only mentioned scalars and vectors.

**Pull request merged**